### PR TITLE
Plotting with SVG

### PR DESCRIPTION
**Pull request by mvollmer**
This is a evolution of #14850.

- [x] Squash all the things.

### CODE REVIEW

**Comment by mvollmer** [comment]
The basics work, the rest is polish. :-)

image

**Comment by mvollmer** [comment]
Alright:

image

All the elements are there, but need professional styling of course.

There is also a lot of computation in JavaScript still.

**Comment by mvollmer** [comment]
This takes nearly no CPU at all. The worst seems to be the construction of the "d" attribute for the `<path>` elements.

**Comment by mvollmer** [comment]
I start to be very happy with this, as a replacement for the current plots.

There is some problem with the interpolation in `cockpit.grid`, and maybe other things:

image

The current plots show the same artifacts. I'll investigate this later on.

**Comment by garrett** [comment]
I rebased this branch and force-pushed, so that it would build again.

**Comment by garrett** [comment]
I auto-fixed eslint issues; there still are 9 ones that cannot be auto-fixed, related to the placement of functions, it looks like.

Additionally, there's a CSS import for a file that doesn't exist, so I removed it.

And now, I have an issue where jQuery doesn't exist.

Screenshot:
Screenshot from 2020-11-19 17-39-27

**Comment by mvollmer** [comment]
> Additionally, there's a CSS import for a file that doesn't exist, so I removed it.

I forgot to commit that file, sorry.  I'll fix up the branch and make it eslint-clean.

**Comment by mvollmer** [comment]
> > Additionally, there's a CSS import for a file that doesn't exist, so I removed it.
> 
> I forgot to commit that file, sorry. I'll fix up the branch and make it eslint-clean.

Done.

**Comment by andreasn** [comment]
Experimented a bit with the colors
https://github.com/andreasn/cockpit/tree/graphtests

image

**Comment by mvollmer** [comment]
Ok, I think I got it.  This now replaces the Storage and Network graphs.  It doesn't yet replace the dashboard, which would allow us to cleanup `pkg/lib/plot.js` drastically. I am in favour of getting this in, and doing the dashboard and cleanup separately.

**Comment by mvollmer** [comment]
> https://github.com/andreasn/cockpit/tree/graphtests

Thanks, included here now.

**Comment by mvollmer** [comment]
@garrett, about switching to lines: Right now the coordinates are for stacked area charts.  That is, in a plot like this:

image

at time 15:36, the red gadget is at about 90 MiB/s and the green at about 30 MiB/s.  Together they are at 120 MiB/s.  So we can't just use the same coordinates for line graphs.

**Comment by garrett** [comment]
@mvollmer and I talked in IRC about this quite a bit. Here are a few highlights:

- Consider using requestAnimationFrame to sync the charts to second https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame
- Perhaps use React hooks with requestAnimationFrame? https://css-tricks.com/using-requestanimationframe-with-react-hooks/
- Additionally, we talked about the Page Visibility API and how that could benefit Cockpit as a whole: https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API (note: browsers should automatically throttle down requestAnimationFrame based on Page Visibility without doing anything w/ the Visibility API, which should at least help for our charts)
- It seems charts currently keep rendering even when the page isn't in the iframe

**Comment by garrett** [comment]
> Right now the coordinates are for stacked area charts.

Aha! I didn't realize that. I'll readjust the commit.

**Comment by mvollmer** [comment]
@garret, I checked and there is nothing wrong about how colors are assigned to the individual instances. The first instance at the bottom will get color 0, the next on top of it gets 1, etc.  If a new instance is added, it gets the next color.  Instances are never removed (unless the page is reloaded) so every instance keeps their color.

I think we can just list the colors in their natural order, with the ones we want to be used first at the start of the array.

**Comment by garrett** [comment]
>  Instances are never removed (unless the page is reloaded) so every instance keeps their color.

Ah, I have a bunch of inactive devices on both pages, so my colors weren't as expected. OK. :+1:

**Comment by garrett** [comment]
More about a11y and SVG, specifically graphs: https://css-tricks.com/accessible-svgs/#complex-images-an-accessible-graph

I'm wondering how we can have a realtime graph actually be useful for screen readers whatsoever.

What we _can_ do is consider keyboard navigation for interacting with the graphs. And make sure they're colorblind accessible (which is what I've already done).

**Comment by mvollmer** [comment]
Selenium tests are looking for the old graphs...

File: pkg/lib/cockpit-components-plot.jsx [hunk 1]
```diff
@@ -411,10 +412,13 @@ export const SvgPlot = ({ title, config, plot_state, plot_id, onHover, className
 
         return (
             <svg width={w} height={h}
+                 ariaLabel={title}
```
Comment (mvollmer) [inline pkg/lib/cockpit-components-plot.jsx:415]: This needs to be `aria-label`.

Comment (garrett) [reply pkg/lib/cockpit-components-plot.jsx:415]: Ah, right. I forgot. React is inconsistent here.

**Comment by martinpitt** [comment]
Not yet ready for review/landing, please click "ready for review" once it is. Thanks!

File: pkg/lib/cockpit-components-plot.jsx [hunk 2]
```diff
@@ -0,0 +1,487 @@
...
+    var n = new Date();
+    var l = new Date(start_ms);
+
+    const year_index = 0;
+    const month_index = 1;
+    const day_index = 2;
+    const hour_minute_index = 3;
+
+    let format_begin;
+    const format_end = hour_minute_index;
```
Comment (mvollmer) [inline pkg/lib/cockpit-components-plot.jsx:81]: We can simplify the code below quite a bit.

File: pkg/lib/cockpit-components-plot.jsx [hunk 3]
```diff
@@ -0,0 +1,487 @@
...
+
+    return {
+        ticks: ticks,
+        formatter: format_tick,
+        start: start_ms,
+        end: end_ms
+    };
+}
+
+function value_ticks(data, config) {
```
Comment (mvollmer) [inline pkg/lib/cockpit-components-plot.jsx:138]: The comments should include a couple of worked examples, I'd say.

File: pkg/lib/plot.js [hunk 4]
```diff
@@ -17,128 +17,34 @@
...
- * channel.
- *
- * The single value for the series is computed by summing the values
- * for all metrics and all instances that are delivered by the
- * channel.
- *
- * The 'desc' argument determines the channel options:
+/* The public API here is not general and a bit weird.  It can only do
+ * what Cockpit itself needs right now, and what can be easily
+ * supported by the old stuff above.  But it's a start.
```
Comment (mvollmer) [inline pkg/lib/plot.js:24]: Let's not talk about "old stuff above".  The docs should be expanded a bit.

File: pkg/networkmanager/interfaces.js [hunk 5]
```diff
@@ -3112,8 +3028,7 @@ PageNetworkInterface.prototype = {
         }
 
         $('#network-interface-members').prop('hidden', true);
-        if (self.main_connection)
```
Comment (mvollmer) [inline pkg/networkmanager/interfaces.js:3115]: This change could be moved into its own commit with a  proper justification.

File: pkg/lib/cockpit-components-plot.jsx (hunk 3 above)
Comment (mvollmer) [reply pkg/lib/cockpit-components-plot.jsx:138]: Done.

File: pkg/lib/cockpit-components-plot.jsx (hunk 2 above)
Comment (mvollmer) [reply pkg/lib/cockpit-components-plot.jsx:81]: Done.

File: pkg/lib/plot.js (hunk 4 above)
Comment (mvollmer) [reply pkg/lib/plot.js:24]: Done.

File: pkg/networkmanager/interfaces.js (hunk 5 above)
Comment (mvollmer) [reply pkg/networkmanager/interfaces.js:3115]: Done.

File: pkg/lib/hooks.js [hunk 6]
```diff
@@ -277,7 +277,7 @@ export function useObject(create, destroy, deps, comps) {
  * }
  *
  * The component will be re-rendered whenever "proxy" emits the
- * "changed" signal.
+ * "changed" signal.  The "proxy" parameter can be null.
```
Comment (KKoukiou) [inline pkg/lib/hooks.js:280]: Is this for safeguarding or what? If obj is null this hook will do nothing really.

File: pkg/lib/cockpit-components-plot.jsx [hunk 7]
```diff
@@ -0,0 +1,521 @@
...
+        return (
+            <svg width={w} height={h}
+                 ariaLabel={title}
+                 role="img"
+                 onMouseDown={plot_state.zoom_state && plot_state.zoom_state.enable_zoom_in ? start_dragging : null}
+                 onMouseUp={selection ? stop_dragging : null}
+                 onMouseMove={selection ? drag : null}
+                 onMouseLeave={cancel_dragging}>
+                <title>{title}</title>
+                <text x={0} y={-20} style={{ fontSize: "small" }} ref={measure_ref}>{config.widest_string}</text>
```
Comment (KKoukiou) [inline pkg/lib/cockpit-components-plot.jsx:450]: Isn't that inline styles which our CSP will block?

File: pkg/lib/cockpit-components-plot.jsx [hunk 8]
```diff
@@ -0,0 +1,521 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2020 Red Hat, Inc.
```
Comment (KKoukiou) [inline pkg/lib/cockpit-components-plot.jsx:4]: Happy new year @mvollmer  :)

**Review by KKoukiou — COMMENTED** [review]
I have some minor comments, everything works as expected otherwise! That was very cool change that you did @mvollmer

File: pkg/lib/cockpit-components-plot.jsx [hunk 9]
```diff
@@ -0,0 +1,521 @@
...
+                            .map((s, j) =>
+                                <tspan key={i + "." + j} x={x_coord(t)} dy="1.2em">{s}</tspan>) }
+                </text>) }
+                { selection &&
+                <rect x={selection.left} y={m_top} width={selection.right - selection.left} height={h - m_top - m_bottom}
+                        stroke="black" fill="tan" opacity="0.5" shapeRendering="crispEdges" /> }
+            </svg>);
+    }
+
+    return (
```
[the remark this answers is not in the document]
Comment (garrett) [reply pkg/lib/cockpit-components-plot.jsx:495]: PF doesn't have a utility class to hide. I think they're relying on authors to use aria-hidden.

Meanwhile, I pushed up a change that hides the plots.

We should re-evaluate a11y after this PR lands. (Not necessarily making the charts accessible, but by providing the essential meaning behind the graphs in an alternate way. One possible idea would be to identify items in the list which have "high" IO/bandwidth somehow.)

File: pkg/lib/hooks.js (hunk 6 above)
Comment (mvollmer) [reply pkg/lib/hooks.js:280]: This is because of the "Rules of  Hooks", https://reactjs.org/docs/hooks-rules.html. Hooks need to be called in the same order for every render.  Thus, this code is not allowed because `useState` might sometimes be called first (when proxy is falsey) and sometimes second (when it is not).

```
function foo (proxy) {
    if (proxy)
        useEvent(proxy);
    const [ value, setValue ] = useState(0);
}
```

The reason for this rule is found in the implementation of hooks, and they break when we violate it.  It's not a style issue.

File: pkg/lib/cockpit-components-plot.jsx (hunk 7 above)
Comment (mvollmer) [reply pkg/lib/cockpit-components-plot.jsx:452]: CSP doesn't apply to React, no?

Comment (KKoukiou) [reply pkg/lib/cockpit-components-plot.jsx:452]: Not directly, but I would expect that the HTML generated from this would also have a 'style' attribute.

File: pkg/lib/cockpit-components-plot.jsx [hunk 10]
```diff
@@ -0,0 +1,521 @@
...
+            size = sizes_in_seconds[i] * 1000;
+            break;
+        }
+    }
+
+    // Determine what to omit from the tick label.  If it's all in the
+    // current year, we don't need to include the year; and if it's
+    // all happening today, we don't include the month and date.
+
+    var n = new Date();
```
[the remark this answers is not in the document]
Comment (mvollmer) [reply pkg/lib/cockpit-components-plot.jsx:73]: Yes, good point. Done.

File: pkg/lib/cockpit-components-plot.jsx [hunk 11]
```diff
@@ -0,0 +1,521 @@
...
+                ]} />
+            { "\n" }
+            <Button variant="secondary" onClick={() => zoom_state.zoom_out()}
+                    isDisabled={!zoom_state.enable_zoom_out}>
+                <span className="glyphicon glyphicon-zoom-out" />
+            </Button>
+            { "\n" }
+            <Button variant="secondary" onClick={() => zoom_state.scroll_left()}
+                    isDisabled={!zoom_state.enable_scroll_left}>
+                <span className="fa fa-angle-left" />
```
[the remark this answers is not in the document]
Comment (mvollmer) [reply pkg/lib/cockpit-components-plot.jsx:281]: Nice, done!

**Comment by mvollmer** [comment]
> I have some minor comments, everything works as expected otherwise! That was very cool change that you did @mvollmer

Thanks, and thanks for the review!

**Comment by garrett** [comment]
I have some changes pending to split the CSS out to a separate file. But I'm waiting on my machine to finish compiling. It feels extra slow today. Anyway, I'll push it up when it's ready. Expect it soon.

File: pkg/lib/cockpit-components-plot.scss [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+    &-axis,
+    &-unit {
+        font-size: var(--pf-global--FontSize--xs);
+        fill: var(--pf-global--Color--200);
+    }
+
+    &-lines,
+    &-ticks {
+        stroke: var(--pf-chart-global--Fill--Color--300, var(--pf-global--Color--light-300));
```
Comment (garrett) [inline pkg/lib/cockpit-components-plot.scss:26]: I was hesitant to make a hard dependency on PF Charts. After looking at the file, I think it's OK, as it just assigns variables (which is what we want).

I can redo the colors to be the chart ones, so our SVG Plots are always up to date with the correct colors, font sizes, set.

Comment (KKoukiou) [reply pkg/lib/cockpit-components-plot.scss:26]: What is the issue on relying on patternfly-charts.scss variables? Or the reasoning behind 'I was hesitant to make a hard dependency on PF Charts.'.

File: pkg/lib/cockpit-components-plot.scss [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+    &-title {
+        font-size: var(--pf-global--FontSize--md);
+    }
+
+    // Placeholder string to stretch the column, set offscreen
+    &-widest {
+        fill: transparent;
+    }
+
+    &-unit {
```
Comment (mvollmer) [inline pkg/lib/cockpit-components-plot.scss:14]: This is duplication below, no?

File: pkg/lib/cockpit-components-plot.scss [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+        fill: var(--pf-global--Color--200);
+    }
+
+    &-axis,
+    &-unit {
```
Comment (mvollmer) [inline pkg/lib/cockpit-components-plot.scss:19]: Duplicated here.

File: pkg/lib/cockpit-components-plot.scss (hunk 13 above)
Comment (garrett) [reply pkg/lib/cockpit-components-plot.scss:14]: Yep. I pushed what I had.

For debugging purposes, I had all of them separate, including the x and y classes. It's fixed locally, with a bunch of other changes too. I'll push up my current changes—with a fix for this—soon.

File: pkg/lib/cockpit-components-plot.jsx [hunk 15]
```diff
@@ -0,0 +1,513 @@
...
+                    range_item(5 * 60, _("5 minutes")),
+                    range_item(60 * 60, _("1 hour")),
+                    range_item(6 * 60 * 60, _("6 hours")),
+                    range_item(24 * 60 * 60, _("1 day")),
+                    range_item(7 * 24 * 60 * 60, _("1 week"))
+                ]} />
+            { "\n" }
+            <Button variant="secondary" onClick={() => zoom_state.zoom_out()}
+                    isDisabled={!zoom_state.enable_zoom_out}>
+                <span className="glyphicon glyphicon-zoom-out" />
```
Comment (KKoukiou) [inline pkg/lib/cockpit-components-plot.jsx:281]: @mvollmer I forgot these here:
Pleese replace with SearchMinusIcon,SearchPlusIcon

That will allow us in the end to remove patternfly-3 and bootstrap from pages importing this component.

File: pkg/lib/cockpit-components-plot.scss [hunk 16]
```diff
@@ -33,4 +51,37 @@
...
+$plotColorCurrent: 0;
+$plotColorTotal: 0;
+
+// Count up total number of plot colors
+@each $plotColor in $plotColors { $plotColorTotal: $plotColorTotal + 1; }
+
+// Iterate through colors and set each graph area to a color
+@each $plotColor, $plotColorBackup in $plotColors {
+    $plotColorCurrent: $plotColorCurrent + 1;
+    .ct-plot-paths > path:nth-child(#{$plotColorTotal}n + #{$plotColorCurrent}) {
```
Comment (mvollmer) [inline pkg/lib/cockpit-components-plot.scss:75]: We need to assign colors in reverse, the last `path` element needs to get the first color from the array, the second last `path` gets the second color from the array.  Otherwise all colors change when a new device is hot-plugged while the page is running.

This might or might not be important.  If a new device is is hot-plugged and then you reload the page, the colors change anyway. (But in a different way than they change before reload with this CSS here.)  It's a bit of a mess.

The `path` elements are in reverse order in the DOM because of the how the data structures with the sample are currently constructed. That can of course be changed, but I think in order for the PR to land without undue delay, we shouldn't attempt that here.

Comment (mvollmer) [reply pkg/lib/cockpit-components-plot.scss:75]: Oh, there is `:nth-last-child`! I thought we had a real problem here. :-)

Comment (mvollmer) [reply pkg/lib/cockpit-components-plot.scss:75]: Done.

File: pkg/lib/cockpit-components-plot.jsx (hunk 15 above)
Comment (mvollmer) [reply pkg/lib/cockpit-components-plot.jsx]: Done.

File: pkg/lib/cockpit-components-plot.scss (hunk 16 above)
Comment (garrett) [reply pkg/lib/cockpit-components-plot.scss]: Wait. nth-last-child does it from reverse... wouldn't it be better to have the elements be in order in the SVG itself instead of doing everything in reverse?

**Comment by garrett** [comment]
LGTM. :+1: I'm happy with merging this and solving anything else in follow-ups.

I did notice that there's extra space at the bottom of the graph for me. Not when it initially loads (just for a flicker), but when there's data. But I'm fine with us figuring this out in a follow-up. We have a week until release.

**Review by garrett — APPROVED** [review]
:+1:

**Comment by mvollmer** [comment]
Test failures are fixed by #15136, I think.  I'll squash and everything will be green...

File: pkg/lib/cockpit-components-plot.jsx [hunk 17]
```diff
@@ -0,0 +1,513 @@
...
+                    </DropdownItem>,
+                    <DropdownSeparator key="sep" />,
+                    range_item(5 * 60, _("5 minutes")),
+                    range_item(60 * 60, _("1 hour")),
+                    range_item(6 * 60 * 60, _("6 hours")),
+                    range_item(24 * 60 * 60, _("1 day")),
+                    range_item(7 * 24 * 60 * 60, _("1 week"))
+                ]} />
+            { "\n" }
+            <Button variant="secondary" onClick={() => zoom_state.zoom_out()}
```
Comment (KKoukiou) [inline pkg/lib/cockpit-components-plot.jsx:279]: Icons which don't have text but  they are wrapping icons should have an aria-label for accessibility. We can fix it in a followup.

**Review by KKoukiou — APPROVED** [review]
Great work @mvollmer!

File: pkg/lib/cockpit-components-plot.jsx (hunk 17 above)
Comment (garrett) [reply pkg/lib/cockpit-components-plot.jsx:279]: These icons should also have a tooltip when hovered too. Everyone should be able to understand the meaning of an icon. (Although a magnifying glass is commonly used for zooming, you'd be surprised at how many people don't understand common icons.)

Comment (garrett) [reply pkg/lib/cockpit-components-plot.jsx:279]: Good catch, BTW!

**Comment by mvollmer** [comment]
`TestMetrics.testCurrentMetrics` fails pretty reliably on rhel-8-4, but I am sure it is actually unrelated to the changes in this PR (despite having to do with metrics as well).
